src/libraries/System.Text.Json/System.Text.Json.SourceGeneration/JsonSerializableSyntaxReceiver.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace System.Text.Json.SourceGeneration
+{
+        public class JsonSerializableSyntaxReceiver : ISyntaxReceiver
+        {
+            public List<KeyValuePair<string, IdentifierNameSyntax>> ExternalClassTypes = new List<KeyValuePair<string, IdentifierNameSyntax>>();
+            public List<KeyValuePair<string, TypeDeclarationSyntax>> InternalClassTypes = new List<KeyValuePair<string, TypeDeclarationSyntax>>();
+
+            public void OnVisitSyntaxNode(SyntaxNode syntaxNode)
+            {
+                // Look for classes or structs for JsonSerializable Attribute.
+                if (syntaxNode is ClassDeclarationSyntax || syntaxNode is StructDeclarationSyntax)
+                {
+                    // Find JsonSerializable Attributes.
+                    IEnumerable<AttributeSyntax>? serializableAttributes = null;
+                    AttributeListSyntax attributeList = ((TypeDeclarationSyntax)syntaxNode).AttributeLists.SingleOrDefault();
+                    if (attributeList != null)
+                    {
+                        serializableAttributes = attributeList.Attributes.Where(node => (node is AttributeSyntax attr && attr.Name.ToString() == "JsonSerializable")).Cast<AttributeSyntax>();
```

Comment:
The trouble with this approach is that you can't guarantee that the attribute name will actually be called `"JsonSerializable"`

For example, here are a bunch of ways I can rename the attribute https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzhgHwAEAmARgFgAoIgBgAIBBAOwgwAsYoA5bAWxi4ADtjAx6AXnpEyAbmp16AZQgCOAS2YBzAKIAbXDB3MM62HoCek6WQB0Sruux71AL2zA9MRhgxR1wACuGDDyVNQA2g7+zm4eXgC6CgDM0iRM9ADeAL7UkTL2jrHunt6+/kEhSTSppPQAQlm54VQRLGycPPyCImKFMS4liSlp9ADCTXmtKmrsmroGRiZmMJbVRLXpACJNQA===

Basically when dealing with syntax, you can only make guesses at things, not full decisions. You won't be able to actually know if the same attribute without doing binding. That's what the compiler does when you ask for a semantic model (it's not a trivial operation ;)) 

The way the other 'attribute based' generators solve this is by just selecting candidates for inclusion (basically anything that has an attribute):
https://github.com/dotnet/roslyn-sdk/blob/94087f6f1d5414ca187d8836caf7c8a114f22978/samples/CSharp/SourceGenerators/SourceGeneratorSamples/AutoNotifyGenerator.cs#L176

Then doing the actual selection during generation. https://github.com/dotnet/roslyn-sdk/blob/94087f6f1d5414ca187d8836caf7c8a114f22978/samples/CSharp/SourceGenerators/SourceGeneratorSamples/AutoNotifyGenerator.cs#L50

As your attribute is provided via references (not injected directly via the generator) you won't need to do the 'create a new compilation' step you'll see above.

Reply:
cc @AaronRobinsonMSFT https://github.com/dotnet/runtimelab/blob/DllImportGenerator/DllImportGenerator/DllImportGenerator/DllImportGenerator.cs#L210 seems to have the same issues.

Reply:
@jkotas Not really. I handle all those cases. I ensure the name ends with `GeneratedDllImportAttribute` or `GeneratedDllImport`. This handles all the cases in that example see https://github.com/dotnet/runtimelab/blob/35c9adc2b099a51aff24992e15380a1f26a60d39/DllImportGenerator/DllImportGenerator/DllImportGenerator.cs#L188-L193.

This doesn't handle the other case of `TrickGeneratedDllImportAttribute` though. Still on the fence whether it is worth addressing based on the cost of building up the semantic model where it is being done.